`config`: add support for `message.timestamp.{before/after}.max.ms`

WillemKauf opened the pull request:
Implements KIP-937 [as described here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-937%3A+Improve+Message+Timestamp+Validation).

Introduces `message.timestamp.{before/after}.max.ms` and `log_message_timestamp_{before/after}_max_ms` at the topic and cluster level.

These can be used to reject records produced with timestamps outside the set bounds by returning [Kafka error code 32](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=255071414#KIP937:ImproveMessageTimestampValidation-PublicInterfaces) (`INVALID_TIMESTAMP`).

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Features

* Adds support for KIP-937 by implementing `message.timestamp.{before/after}.max.ms`.
* Deprecates `log_message_timestamp_alert_{before/after}_ms` cluster properties.

rockwotj committed:
kafka: set max_timestamp if missing and using timestamp_type::create_time

Some clients (looking at you Sarama!) don't set a max_timestamp for
batches when being produced. In Apache Kafka, all incoming batches have
the max_timestamp batch header field set BY THE BROKER. Code references:

Here for uncompressed batches:
https://github.com/apache/kafka/blob/e124d3975bdb3a9ec85eee2fba7a1b0a6967d3a6/storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java#L275

Here for compressed batches:
https://github.com/apache/kafka/blob/e124d3975bdb3a9ec85eee2fba7a1b0a6967d3a6/storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java#L404

When the timestamp_type is log_append then we do set the max_timestamp
in the broker:
https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/src/v/kafka/server/handlers/produce.cc#L309

But don't do anything in the create_time case. This has the effect of
breaking timequeries, which only look at the max_timestamp in many
places of the storage layer (which is similar to the Kafka behavior, but
they have the max_timestamp set correctly).

However historically we've not decompressed batches on the produce path
(which is technically a validation bug). Since that has many perf
impliciations and doing that decompressing would presumably need at
least a cluster config flag, we only partially fix this bug when the
batch is uncompressed. For compressed batches we should really just
do the right thing and validate the batch (but maybe behind a flag?).

WillemKauf committed:
`config`: add `message_timestamp_{before/after}_max_ms`

WillemKauf committed:
`cluster`: add `message_timestamp_{before/after}_max_ms` to `topic_configuration`

WillemKauf committed:
`kafka`: add `message_timestamp_{before/after}_max_ms` to topic config handlers

WillemKauf committed:
`config`: deprecate `log_message_timestamp_alert_{before/after}_ms`

We have topic properties that will actively reject messages now,
so these alerts are un-needed and would only serve to confuse users
due to a mismatched configuration between the alert and the other
properties which reject the records.

WillemKauf committed:
`kafka`: reject messages based on `message_timestamp_{before/after}_max_ms`

This commit properly enforces these configs within the produce path.

There is also a subtle reworking of `validate_batch_timestamps()`,
though there are no other functional changes in the reworking.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71583-->
<!--53c61858-4b11-4078-a7a3-3cf11348f2cc-->
<!--6f95bb1a-a2e6-47cc-b78b-f6c345120958-->
<!--992a2865-7e01-4ec0-a58c-a8ccce42f61e-->
<!--a41cda5d-9c64-4513-8c00-628749404131-->
<!--fd8f8579-e71c-4575-80b9-7a455c67873f-->
<!--5f5e2382-6753-4a51-b402-354bf27ea646-->
<!--ff45936f-860b-4e70-b2aa-6529310fc61b-->
<!--bc5846ee-dfd6-4e61-b586-a42425a0096e-->
<!--title-->
## Retry command for [Build#71583](https://buildkite.com/redpanda/redpanda/builds/71583)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/retention_policy_test.py::BogusTimestampTest.test_bogus_timestamps@{"mixed_timestamps":true,"use_broker_timestamps":false}
tests/rptest/tests/retention_policy_test.py::BogusTimestampTest.test_bogus_timestamps@{"mixed_timestamps":true,"use_broker_timestamps":true}
tests/rptest/tests/retention_policy_test.py::BogusTimestampTest.test_bogus_timestamps@{"mixed_timestamps":false,"use_broker_timestamps":false}
tests/rptest/tests/retention_policy_test.py::BogusTimestampTest.test_bogus_timestamps@{"mixed_timestamps":false,"use_broker_timestamps":true}
tests/rptest/tests/compatibility/kafka_streams_test.py::KafkaStreamsSessionWindow.test_kafka_streams
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5047a7ed-f4f0-489c-a6bc-b1dcab8aaf91-->
<!--TEST RESULTS-->
<!--274ff3c2-cb85-4663-b7a5-34e67a8020f2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71583">build#71583</a></summary>


| test_class                | test_method           | test_arguments                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:--------------------------|:----------------------|:------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| KafkaStreamsSessionWindow | test_kafka_streams    | null                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cb-96c5-4af5-9987-524f78dd0b93 | FAIL          | 0/21     | The test has failed across all retries |
| KafkaStreamsSessionWindow | test_kafka_streams    | null                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cd-2b65-4142-b267-4ab9a2f78c94 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": false, "use_broker_timestamps": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cb-96c4-4b4d-97d3-f1b6294dbddb | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": false, "use_broker_timestamps": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cd-2b65-4b54-8361-feb77bda5753 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": false, "use_broker_timestamps": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cb-96c5-4af5-9987-524f78dd0b93 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": false, "use_broker_timestamps": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cd-2b65-4142-b267-4ab9a2f78c94 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": true, "use_broker_timestamps": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cb-96be-4c38-9586-77501cea24d5 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": true, "use_broker_timestamps": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cd-2b5f-4696-bbd0-e2789096dec7 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": true, "use_broker_timestamps": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cb-96bf-4220-b0da-3eb6987c0546 | FAIL          | 0/21     | The test has failed across all retries |
| BogusTimestampTest        | test_bogus_timestamps | {"mixed_timestamps": true, "use_broker_timestamps": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/71583#0198f3cd-2b60-4e1d-b037-f39be9a5a603 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71645">build#71645</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71645#0198f7dd-7cd2-4cb0-be78-28f42a0c41aa | FLAKY         | 20/21    | upstream reliability is '99.13419913419914'. current run reliability is '95.23809523809523'. drift is 3.8961 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -201,27 +203,98 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
-        return std::nullopt;
+        batch.set_max_timestamp(
+          model::timestamp_type::append_time, broker_time);
+    }
```
Missing `return`

WillemKauf replied:
Oops, this disappeared, thanks

rockwotj commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -201,27 +203,98 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
+    // `message.timestamp_type` is `CreateTime`. Validate record timestamps
+    // using `message.timestamp.before.max.ms` and
+    // `message.timestamp.after.max.ms`.
+    auto broker_timepoint = model::duration_since_epoch(broker_time);
+
+    // reject if first_timestamp is too far in the past
+    auto first_timepoint = model::duration_since_epoch(header.first_timestamp);
+    if (
+      broker_timepoint > first_timepoint
+      && std::chrono::duration_cast<std::chrono::milliseconds>(
+           broker_timepoint - first_timepoint)
+           > message_timestamp_before_max_ms) {
```
Remind me why this check can't be:

```
auto min_timepoint = broker_time - message_timestamp_before_max_ms;
if (first_timestamp < min_timepoint) {
  // error
}
```

Is it something about overflow?

WillemKauf replied:
Yes, `message_timestamp_before_max_ms` is a bounded property that defaults to `serde::max_serializable_ms` (`9223372036854`), so subtracting this value from broker time would be an underflow (until `Friday, April 11, 2262 11:47:16.854 PM` :smile: )

~~The checks here are safe to under/overflow in their current implementation.~~

rockwotj replied:
ugh I miss `absl::Time` that has saturating logic instead...

rockwotj commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -201,27 +203,98 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
+      broker_timepoint < max_timepoint
+      && std::chrono::duration_cast<std::chrono::milliseconds>(
+           max_timepoint - broker_timepoint)
+           > message_timestamp_after_max_ms) {
```
ditto, why can't this be:

```
auto max_timepoint = broker_timepoint + message_timestamp_after_max_ms;
if (header_max_timestamp > max_timepoint) {
  // error
}
``

WillemKauf replied:
Potential overflow, better to subtract.

rockwotj reviewed (commented):
Nice, just a few small bits on the validation itself (also CI failures)

rockwotj reviewed (dismissed):
LGTM

rockwotj commented:
Also please mention the property deprecation in the release notes

WillemKauf commented:
> Also please mention the property deprecation in the release notes

Done, i'll wait for a docs review of the cluster property description as well before merging

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -825,27 +825,37 @@ configuration::configuration()
+      "difference will be rejected. This property is ignored for "
+      "`message.timestamp.type` set to `AppendTime`. The topic property "
+      "`message.timestamp.before.max.ms` overrides the value of "
+      "`log_message_timestamp_before_max_ms` at the topic level.",
```
Suggested text:

> The maximum allowable timestamp difference between the broker's timestamp and a record's timestamp. For topics with `message.timestamp.type` set to `CreateTime`, Redpanda rejects records that have timestamps earlier than the broker timestamp and exceed this difference. Redpanda ignores this property for topics with `message.timestamp.type` set to `AppendTime`. The topic property `message.timestamp.before.max.ms` overrides the value of `log_message_timestamp_before_max_ms` at the topic level.


```suggestion
The maximum allowable timestamp difference between the broker's timestamp and a record's timestamp. For topics with `message.timestamp.type` set to `CreateTime`, Redpanda rejects records that have timestamps earlier than the broker timestamp and exceed this difference. Redpanda ignores this property for topics with `message.timestamp.type` set to `AppendTime`. The topic property `message.timestamp.before.max.ms` overrides the value of `log_message_timestamp_before_max_ms` at the topic level.
```

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -825,27 +825,37 @@ configuration::configuration()
+  , log_message_timestamp_after_max_ms(
+      *this,
+      "log_message_timestamp_after_max_ms",
+      "The maximum allowable timestamp difference between the broker's "
+      "timestamp and a record's timestamp. For topics with "
+      "`message.timestamp.type` set to `CreateTime`, records which have a "
+      "timestamp later than the broker timestamp that exceed this "
+      "difference will be rejected. This property is ignored for "
+      "`message.timestamp.type` set to `AppendTime`. The topic property "
+      "`message.timestamp.after.max.ms` overrides the value of "
+      "`log_message_timestamp_after_max_ms` at the topic level.",
```
Suggested text (similar as before: 

> The maximum allowable timestamp difference between the broker's timestamp and a record's timestamp. For topics with `message.timestamp.type` set to `CreateTime`, Redpanda rejects records that have timestamps later than the broker timestamp and exceed this difference. Redpanda ignores this property for topics with `message.timestamp.type` set to `AppendTime`. The topic property `message.timestamp.after.max.ms` overrides the value of `log_message_timestamp_after_max_ms` at the topic level.


```suggestion
The maximum allowable timestamp difference between the broker's timestamp and a record's timestamp. For topics with `message.timestamp.type` set to `CreateTime`, Redpanda rejects records that have timestamps later than the broker timestamp and exceed this difference. Redpanda ignores this property for topics with `message.timestamp.type` set to `AppendTime`. The topic property `message.timestamp.after.max.ms` overrides the value of `log_message_timestamp_after_max_ms` at the topic level.
```

WillemKauf committed:
`rptest`: add `timestamp_policy_test`

A variety of test cases that produce records with timestamps in the past,
future, and present, to assert that `message.timestamp.{before/after}.max.ms`
are being correctly enforced.

WillemKauf committed:
`kafka`: improve over/underflow handling in `validate_batch_timestamps()`

Kafka records can be produced with timestamps up to `int64_t::max()`.

Before, validation checks for timestamps in the produce path were
implemented via the following:

```
auto broker_time = model::timestamp::now();
auto broker_timepoint = model::duration_since_epoch(broker_time);
auto max_timepoint = model::duration_since_epoch(header.max_timestamp);
if (broker_timepoint < max_timepoint
  && std::chrono::duration_cast<std::chrono::milliseconds>(
       max_timepoint - broker_timepoint)
       > message_timestamp_after_max_ms) {
   ...
}
```

where `model::duration_since_epoch()` is defined as:

```
inline timestamp_clock::duration duration_since_epoch(timestamp ts) {
    return std::chrono::duration_cast<timestamp_clock::duration>(
      std::chrono::milliseconds{ts.value()});
}
```

where `timestamp_clock::duration` is `std::chrono::system_clock::duration`,
which can be `microseconds` or `nanoseconds`. Uh oh.

So, in the case that the `max_timepoint` is near `int64_t::max()`, we
effectively `duration_cast` a `std::chrono::milliseconds` type to a
`std::chrono::microseconds` type (multiplying underlying value by 1000),
subtract two values, and then attempt to cast back to a
`std::chrono::milliseconds` type, which is only guaranteed to be a signed
integer of at least 45 bits [1].

This can lead to overflow. Fix the issue by keeping everything as a
`model::timestamp`, with `int64_t` comparisions and no opportunity
for narrowing conversions.

[1]:

https://en.cppreference.com/w/cpp/chrono/duration.html

WillemKauf commented:
Force push to: 
* Fix cluster property description per docs review
* Add a new commit that better handles under/overflow in `validate_batch_timestamps()`

@rockwotj please take a look, I can squash the commit in with the original change if you'd like

WillemKauf commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -229,15 +229,12 @@ auto validate_batch_timestamps(
-           > message_timestamp_before_max_ms) {
+      broker_time > header.first_timestamp
+      && (broker_time - header.first_timestamp)
+           > model::timestamp(message_timestamp_before_max_ms.count())) {
```
this is now an `int64_t` comparison. Using `.count()` is fine since `model::timestamp` uses `int64_t` and expects milliseconds.

WillemKauf replied:
`duration_since_epoch()` is a pretty dangerous function, I think.

rockwotj replied:
Agreed, I am actually going to remove it after this PR goes in.

rockwotj commented on src/v/kafka/server/handlers/produce.cc:
Any tests we should add for 99c7a2ba7b0d214b326e217d8c7bdc934c7badd2?

WillemKauf replied:
I'll think about adding some lower level kafka tests here, it would be easier to ensure we aren't getting overflows in a fixture test than in Ducktape. I'll look at a quick follow-up here.

rockwotj reviewed (commented):
LGTM

rockwotj reviewed (approved):
LGTM
